changelogs/unreleased/linearizable-read.md

```diff
@@ -0,0 +1,5 @@
+## feature/core
+
+* Introduced linearizable transactions: turned on by `box.begin{linearizable =
```

Comment:
> Anyway, since it requires read-confirmed (strictest so far) isolation level to be set (according to the doc request), why not introduce a new isolation level instead - isolation_level = 'linearizable'?

I thought about it as well, but decided that 'linearizable' is not an isolation level, after all.
IMO isolation levels are about how transactions performed on one instance see each other's changes, while linearizability is about seeing changes performed on other instances.

Anyway, I'm not against it, but I think I'll have to ask products about that.

Reply:
> In the doc bot request the option is named linearizable_read.

Thanks for noticing! Fixed.

Reply:
> why not introduce a new isolation level instead - isolation_level = 'linearizable'?

Agreed and done:
<details><summary>diff</summary>

```diff
diff --git a/extra/exports b/extra/exports
index 83d08f7a2..607c2fe3d 100644
--- a/extra/exports
+++ b/extra/exports
@@ -138,7 +138,6 @@ box_txn_rollback_to_savepoint
 box_txn_savepoint
 box_txn_set_isolation
 box_txn_set_timeout
-box_txn_make_linearizable
 box_update
 box_upsert
 clock_monotonic
diff --git a/src/box/box.cc b/src/box/box.cc
index 317dce225..c79efd6e9 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -1377,6 +1377,12 @@ box_check_txn_isolation(void)
 			 "to 'default'");
 		return txn_isolation_level_MAX;
 	}
+	if (level == TXN_ISOLATION_LINEARIZABLE) {
+		diag_set(ClientError, ER_CFG, "txn_isolation",
+			 "cannot set default transaction isolation "
+			 "to 'linearizable'");
+		return txn_isolation_level_MAX;
+	}
 	return (enum txn_isolation_level)level;
 }
 
diff --git a/src/box/iproto.cc b/src/box/iproto.cc
index ef544bc9f..4dec89211 100644
--- a/src/box/iproto.cc
+++ b/src/box/iproto.cc
@@ -1977,12 +1977,6 @@ tx_process_begin(struct cmsg *m)
 		(void)rc;
 		goto error;
 	}
-	if (msg->begin.is_linearizable && box_txn_make_linearizable() != 0) {
-		int rc = box_txn_rollback();
-		assert(rc == 0);
-		(void)rc;
-		goto error;
-	}
 	out = msg->connection->tx.p_obuf;
 	header = obuf_create_svp(out);
 	iproto_reply_ok(out, msg->header.sync, ::schema_version);
diff --git a/src/box/lua/schema.lua b/src/box/lua/schema.lua
index 33e143c90..f5f045a68 100644
--- a/src/box/lua/schema.lua
+++ b/src/box/lua/schema.lua
@@ -76,8 +76,6 @@ ffi.cdef[[
     box_txn_set_timeout(double timeout);
     int
     box_txn_set_isolation(uint32_t level);
-    int
-    box_txn_make_linearizable(void);
     void
     memtx_tx_story_gc_step();
     /** \endcond public */
@@ -370,6 +368,8 @@ box.txn_isolation_level = {
     ['READ_CONFIRMED'] = 2,
     ['best-effort'] = 3,
     ['BEST_EFFORT'] = 3,
+    ['linearizable'] = 4,
+    ['LINEARIZABLE'] = 4,
 }
 
 -- Create private isolation level map anything-correct -> number.
@@ -388,6 +388,9 @@ box.internal.txn_isolation_level_map = txn_isolation_level_map
 
 -- Convert to numeric the value of txn isolation level, raise if failed.
 local function normalize_txn_isolation_level(txn_isolation)
+    if txn_isolation == nil then
+        return 0
+    end
     txn_isolation = txn_isolation_level_map[txn_isolation]
     if txn_isolation == nil then
         box.error(box.error.ILLEGAL_PARAMS,
@@ -408,19 +411,15 @@ local begin_options = {
         return true
     end,
     txn_isolation = normalize_txn_isolation_level,
-    linearizable = 'boolean',
 }
 
 box.begin = function(options)
     check_param_table(options, begin_options)
     local timeout
     local txn_isolation
-    local linearizable
     if options then
         timeout = options.timeout
-        txn_isolation = options.txn_isolation and
-                        normalize_txn_isolation_level(options.txn_isolation)
-        linearizable = options.linearizable
+        txn_isolation = options.txn_isolation
     end
     if builtin.box_txn_begin() == -1 then
         box.error()
@@ -428,12 +427,9 @@ box.begin = function(options)
     if timeout then
         assert(builtin.box_txn_set_timeout(timeout) == 0)
     end
-    if txn_isolation and
-       builtin.box_txn_set_isolation(txn_isolation) ~= 0 then
-        box.rollback()
-        box.error()
-    end
-    if linearizable and builtin.box_txn_make_linearizable() ~= 0 then
+    txn_isolation = normalize_txn_isolation_level(txn_isolation)
+    assert(txn_isolation ~= nil)
+    if builtin.box_txn_set_isolation(txn_isolation) ~= 0 then
         box.rollback()
         box.error()
     end
diff --git a/src/box/memtx_tx.c b/src/box/memtx_tx.c
index bb0db2a20..432d0e7ad 100644
--- a/src/box/memtx_tx.c
+++ b/src/box/memtx_tx.c
@@ -2415,7 +2415,8 @@ detect_whether_prepared_ok(struct txn *txn)
 		return false;
 	else if (txn->isolation == TXN_ISOLATION_READ_COMMITTED)
 		return true;
-	else if (txn->isolation == TXN_ISOLATION_READ_CONFIRMED)
+	else if (txn->isolation == TXN_ISOLATION_READ_CONFIRMED ||
+		 txn->isolation == TXN_ISOLATION_LINEARIZABLE)
 		return false;
 	assert(txn->isolation == TXN_ISOLATION_BEST_EFFORT);
 	/*
diff --git a/src/box/txn.c b/src/box/txn.c
index cee5d3feb..2eb6046e2 100644
--- a/src/box/txn.c
+++ b/src/box/txn.c
@@ -54,6 +54,7 @@ const char *txn_isolation_level_strs[txn_isolation_level_MAX] = {
 	"READ_COMMITTED",
 	"READ_CONFIRMED",
 	"BEST_EFFORT",
+	"LINEARIZABLE",
 };
 
 const char *txn_isolation_level_aliases[txn_isolation_level_MAX] = {
@@ -61,6 +62,7 @@ const char *txn_isolation_level_aliases[txn_isolation_level_MAX] = {
 	"read-committed",
 	"read-confirmed",
 	"best-effort",
+	"linearizable"
 };
 
 /* Txn cache. */
@@ -607,7 +609,7 @@ txn_begin_stmt(struct txn *txn, struct space *space, uint16_t type)
 		txn_set_flags(txn, TXN_IS_CONFLICTED);
 	}
 
-	if (space != NULL && txn_has_flag(txn, TXN_IS_LINEARIZABLE) &&
+	if (space != NULL && txn_is_linearizable(txn) &&
 	    txn_check_space_linearizability(txn, space) != 0) {
 		return -1;
 	}
@@ -1370,6 +1372,18 @@ box_txn_set_timeout(double timeout)
 	return 0;
 }
 
+/** Wait for a linearization point for a transaction. */
+static int
+txn_make_linearizable(struct txn *txn)
+{
+	if (!memtx_tx_manager_use_mvcc_engine) {
+		diag_set(ClientError, ER_UNSUPPORTED, "Linearizable "
+			 "transaction", "disabled memtx mvcc engine");
+		return -1;
+	}
+	return box_wait_linearization_point(txn->timeout);
+}
+
 int
 box_txn_set_isolation(uint32_t level)
 {
@@ -1390,35 +1404,8 @@ box_txn_set_isolation(uint32_t level)
 	if (level == TXN_ISOLATION_DEFAULT)
 		level = txn_default_isolation;
 	txn->isolation = level;
-	return 0;
-}
-
-int
-box_txn_make_linearizable(void)
-{
-	struct txn *txn = in_txn();
-	if (txn == NULL) {
-		diag_set(ClientError, ER_NO_TRANSACTION);
-		return -1;
-	}
-	if (!stailq_empty(&txn->stmts)) {
-		diag_set(ClientError, ER_ACTIVE_TRANSACTION);
-		return -1;
-	}
-	if (txn->isolation != TXN_ISOLATION_READ_CONFIRMED) {
-		diag_set(ClientError, ER_UNSUPPORTED,
-			 "Linearizable transaction",
-			 "anything but read-confirmed isolation");
-		return -1;
-	}
-	if (!memtx_tx_manager_use_mvcc_engine) {
-		diag_set(ClientError, ER_UNSUPPORTED, "Linearizable "
-			 "transaction", "disabled memtx mvcc engine");
-		return -1;
-	}
-	if (box_wait_linearization_point(txn->timeout) != 0)
-		return -1;
-	txn_set_flags(txn, TXN_IS_LINEARIZABLE);
+	if (level == TXN_ISOLATION_LINEARIZABLE)
+		return txn_make_linearizable(txn);
 	return 0;
 }
 
diff --git a/src/box/txn.h b/src/box/txn.h
index df662e1cb..2a2fd053d 100644
--- a/src/box/txn.h
+++ b/src/box/txn.h
@@ -102,11 +102,6 @@ enum txn_flag {
 	 * rolled back at commit.
 	 */
 	TXN_IS_ABORTED_BY_TIMEOUT = 0x100,
-	/**
-	 * Transaction is linearizable, so it can only touch certain types of
-	 * spaces (synchronous ones, for example).
-	 */
-	TXN_IS_LINEARIZABLE = 0x200,
 };
 
 enum {
@@ -175,6 +170,8 @@ enum txn_isolation_level {
 	TXN_ISOLATION_READ_CONFIRMED,
 	/** Determine isolation level automatically. */
 	TXN_ISOLATION_BEST_EFFORT,
+	/** Allow to read only the changes confirmed on any cluster node. */
+	TXN_ISOLATION_LINEARIZABLE,
 	/** Upper bound of valid values. */
 	txn_isolation_level_MAX,
 };
@@ -783,6 +780,12 @@ txn_begin_stmt(struct txn *txn, struct space *space, uint16_t type);
 int
 txn_begin_in_engine(struct engine *engine, struct txn *txn);
 
+static inline bool
+txn_is_linearizable(const struct txn *txn)
+{
+	return txn->isolation == TXN_ISOLATION_LINEARIZABLE;
+}
+
 /**
  * Check if a space supports linearizable access, i.e. it is either synchronous
  * or not replicated at all, and its engine prevents dirty reads.
@@ -791,7 +794,7 @@ static inline int
 txn_check_space_linearizability(const struct txn *txn,
 				const struct space *space)
 {
-	assert(txn_has_flag(txn, TXN_IS_LINEARIZABLE));
+	assert(txn_is_linearizable(txn));
 	(void)txn;
 	/*
 	 * XXX: can't promise the user linearizability with Vinyl until
@@ -842,7 +845,7 @@ txn_begin_ro_stmt(struct space *space, struct txn **txn,
 			diag_set(ClientError, ER_TRANSACTION_CONFLICT);
 			return -1;
 		}
-		if (txn_has_flag(*txn, TXN_IS_LINEARIZABLE) &&
+		if (txn_is_linearizable(*txn) &&
 		    txn_check_space_linearizability(*txn, space) != 0) {
 			return -1;
 		}
@@ -1040,14 +1043,6 @@ box_txn_set_timeout(double timeout);
 API_EXPORT int
 box_txn_set_isolation(uint32_t level);
 
-/**
- * Make the transaction linearizable.
- * @retval  0 on success
- * @retval -1 on failure, diag is set.
- */
-API_EXPORT int
-box_txn_make_linearizable(void);
-
 /** \endcond public */
 
 typedef struct txn_savepoint box_txn_savepoint_t;
diff --git a/test/replication-luatest/linearizable_read_test.lua b/test/replication-luatest/linearizable_test.lua
similarity index 90%
rename from test/replication-luatest/linearizable_read_test.lua
rename to test/replication-luatest/linearizable_test.lua
index e6373dd0e..a342e1a4b 100644
--- a/test/replication-luatest/linearizable_read_test.lua
+++ b/test/replication-luatest/linearizable_test.lua
@@ -23,7 +23,6 @@ g.before_all(function()
     g.box_cfg = {
         replication = build_replication(num_servers),
         replication_timeout = 0.1,
-        txn_isolation = 'read-confirmed',
         memtx_use_mvcc_engine = true,
     }
     g.servers[1] = g.cluster:build_and_add_server({
@@ -68,7 +67,7 @@ g.test_wait_others = function(g)
         local t = require('luatest')
         local fiber = require('fiber')
         local f = fiber.new(function()
-            return pcall(box.begin, {linearizable = true})
+            return pcall(box.begin, {txn_isolation = 'linearizable'})
         end)
         f:set_joinable(true)
         fiber.sleep(0.01)
@@ -89,7 +88,8 @@ g.test_timeout = function(g)
     g.servers[1]:exec(function()
         local t = require('luatest')
         box.error.injection.set('ERRINJ_RELAY_FROM_TX_DELAY', true)
-        local ok, err = pcall(box.begin, {linearizable = true, timeout = 0.01})
+        local ok, err = pcall(box.begin, {txn_isolation = 'linearizable',
+                                          timeout = 0.01})
         box.error.injection.set('ERRINJ_RELAY_FROM_TX_DELAY', false)
         t.assert(not ok, 'Error when relay is unresponsive')
         t.assert_equals(err.type, 'TimedOut',
@@ -138,7 +138,8 @@ g.test_no_dirty_reads = function(g)
     end)
     g.servers[1]:exec(function()
         local t = require('luatest')
-        local ok, err = pcall(box.begin, {linearizable = true, timeout = 0.1})
+        local ok, err = pcall(box.begin, {txn_isolation = 'linearizable',
+                                          timeout = 0.1})
         t.assert(not ok, 'Error with unconfirmed transaction')
         t.assert_equals(err.type, 'TimedOut', 'Timeout waiting for confirm')
         t.assert_equals(box.info.synchro.queue.len, 1,
@@ -150,7 +151,7 @@ g.test_no_dirty_reads = function(g)
     end, {g.replication_synchro_quorum})
     g.servers[1]:exec(function()
         local t = require('luatest')
-        box.begin{linearizable = true}
+        box.begin{txn_isolation = 'linearizable'}
         local len = box.info.synchro.queue.len
         local get = box.space.sync:get{1}
         box.commit()
@@ -171,7 +172,7 @@ g.test_reconnect_while_waiting = function(g)
         end)
     end
     local f = fiber.new(g.servers[1].exec, g.servers[1], function()
-        local ok, err = pcall(box.begin, {linearizable = true})
+        local ok, err = pcall(box.begin, {txn_isolation = 'linearizable'})
         if ok then
             box.commit()
         end
@@ -203,7 +204,7 @@ g.test_leader_change = function(g)
     end)
     local fid = g.servers[1]:exec(function()
         local f = require('fiber').new(function()
-            box.begin{linearizable = true}
+            box.begin{txn_isolation = 'linearizable'}
             local owner = box.info.synchro.queue.owner
             local get = box.space.sync:get{2}
             box.commit()
@@ -261,14 +262,14 @@ local function begin(opts)
 end
 
 local function linearizable_read_from_space(spacename)
-    box.begin{linearizable = true, txn_isolation = 'read-confirmed'}
+    box.begin{txn_isolation = 'linearizable'}
     local ok, err = pcall(box.space[spacename].select, box.space[spacename])
     box.commit()
     return ok, err
 end
 
 local function linearizable_write_to_space(spacename)
-    box.begin{linearizable = true, txn_isolation = 'read-confirmed'}
+    box.begin{txn_isolation = 'linearizable'}
     local ok, err = pcall(box.space[spacename].replace,
                           box.space[spacename], {1})
     box.commit()
@@ -291,13 +292,13 @@ local function test_linearizable_access(node, spacename, is_ok, msg)
 end
 
 g_basic.test_basic = function(g)
-    local ok, err = g.nomvcc:exec(begin, {{linearizable = true}})
-    t.assert(not ok, 'Error without read-confirmed isolatoin')
-    t.assert_equals(err.message, 'Linearizable transaction does not support ' ..
-                                 'anything but read-confirmed isolation',
-                    'Correct error without read-confirmed isolation')
-    ok, err = g.nomvcc:exec(begin, {{linearizable = true,
-                                     txn_isolation = 'read-confirmed'}})
+    local ok, err = pcall(box.cfg, {txn_isolation = 'linearizable'})
+    t.assert(not ok, 'Cannot set default isolation level to linearizable')
+    local msg = "Incorrect value for option 'txn_isolation': cannot set " ..
+                "default transaction isolation to 'linearizable'"
+    t.assert_equals(err.message, msg,
+                    'Correct error when trying to set default isolation')
+    ok, err = g.nomvcc:exec(begin, {{txn_isolation = 'linearizable'}})
     t.assert(not ok, 'Error without mvcc enabled')
     t.assert_equals(err.message, 'Linearizable transaction does not support ' ..
                                  'disabled memtx mvcc engine',

```

</details>